File: Dynamic Programming/Number of Ways to Paint N × 3 Grid/solution.ts
// Time complexity: n * 3^3 * 3^3
// Space complexity: n * 2^6 + log(n)
function numOfWays(n: number): number {
    let dp = Array(n).fill(0).map(() => Array(1 << (3 * 2)))
    let colors = [1, 2, 3]
    let MOD = 1e9 + 7
    
    let dfs = (row: number, col: number, previousRow: number, currentRow: number) => {
        if(row >= n) return 1
        if(col >= 3) return dfs(row + 1, 0, currentRow, 0)
        if(col === 0 && dp[row][previousRow] !== undefined) return dp[row][previousRow]
        let result = 0
        let top = (previousRow >> (col * 2)) & 0b11
        let left = (currentRow >> ((col-1) * 2)) & 0b11
        let original = currentRow
        for(let color of colors) {
            if(color === left || color === top) continue
            currentRow |= (color << (col * 2))
            result += dfs(row, col + 1, previousRow, currentRow)
            result %= MOD
            currentRow = original
        }
        if(col === 0) dp[row][previousRow] = result
        return result
    }
    
    return dfs(0, 0, 0, 0)
};

File: Dynamic Programming/Number of Ways to Paint N × 3 Grid/solution.py
class Solution:
    def numOfWays(self, n: int) -> int:
        colors = [1,2,3]
        dp = [[None for j in range(1 << 6)] for i in range(n)]
        MOD = 1e9 + 7
        
        def dfs(row, col, prevMask, currentMask):
            if row >= n:
                return 1
            if col >= 3:
                return dfs(row + 1, 0, currentMask, 0)
            if dp[row][prevMask] != None and col == 0:
                return dp[row][prevMask]
            top = (prevMask >> (2 * col)) & 0b11
            if col > 0:
                left = (currentMask >> (2 * (col - 1))) & 0b11
            else:
                left = 0
            result = 0
            for color in colors:
                if color == top or color == left:
                    continue
                newMask = currentMask | (color << (2*col))
                result += dfs(row, col + 1, prevMask, newMask)
                result = int(result % MOD)
            if col == 0:
                dp[row][prevMask] = result
            return result
        
        return dfs(0, 0, 0, 0)